File: mvp/frontend/lib/timezone.ts
/**
 * Centralized timezone utilities for CommonGround frontend.
 *
 * Architecture:
 * - Backend stores/returns naive UTC datetimes with "Z" suffix
 * - Frontend converts to/from user's profile timezone for display
 * - User's timezone is stored in their profile
 */

import { format, parseISO } from "date-fns";
import { toZonedTime, fromZonedTime } from "date-fns-tz";

/**
 * Supported US timezones
 */
export const TIMEZONE_OPTIONS = [
  { value: "America/New_York", label: "Eastern Time (ET)" },
  { value: "America/Chicago", label: "Central Time (CT)" },
  { value: "America/Denver", label: "Mountain Time (MT)" },
  { value: "America/Los_Angeles", label: "Pacific Time (PT)" },
  { value: "America/Anchorage", label: "Alaska Time (AKT)" },
  { value: "Pacific/Honolulu", label: "Hawaii Time (HT)" },
] as const;

export type TimezoneValue = (typeof TIMEZONE_OPTIONS)[number]["value"];

export const DEFAULT_TIMEZONE: TimezoneValue = "America/Los_Angeles";

/**
 * Get short timezone abbreviation for display
 */
export function getTimezoneAbbr(timezone: string): string {
  const abbrevMap: Record<string, string> = {
    "America/New_York": "ET",
    "America/Chicago": "CT",
    "America/Denver": "MT",
    "America/Los_Angeles": "PT",
    "America/Anchorage": "AKT",
    "Pacific/Honolulu": "HT",
  };
  return abbrevMap[timezone] || timezone;
}

/**
 * Format a UTC datetime string for display in the user's timezone.
 *
 * @param utcString - ISO 8601 UTC datetime string (e.g., "2024-01-15T02:30:00Z")
 * @param timezone - IANA timezone identifier (e.g., "America/Los_Angeles")
 * @param formatStr - date-fns format string
 * @returns Formatted string in user's local time
 *
 * @example
 * formatInUserTimezone("2024-01-16T02:30:00Z", "America/Los_Angeles", "PPp")
 * // Returns: "Jan 15, 2024, 6:30 PM"
 */
export function formatInUserTimezone(
  utcString: string | null | undefined,
  timezone: string,
  formatStr: string = "PPp"
): string {
  if (!utcString) return "";

  try {
    const utcDate = parseISO(utcString);
    const zonedDate = toZonedTime(utcDate, timezone);
    return format(zonedDate, formatStr);
  } catch (error) {
    console.error("Error formatting date:", error);
    return utcString;
  }
}

/**
 * Format time for schedule display with timezone indicator.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - IANA timezone identifier
 * @returns Formatted string like "6:30 PM PT"
 *
 * @example
 * formatScheduleTime("2024-01-16T02:30:00Z", "America/Los_Angeles")
 * // Returns: "6:30 PM PT"
 */
export function formatScheduleTime(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    const timeStr = formatInUserTimezone(utcString, timezone, "h:mm a");
    const tzAbbr = getTimezoneAbbr(timezone);
    return `${timeStr} ${tzAbbr}`;
  } catch (error) {
    console.error("Error formatting schedule time:", error);
    return utcString;
  }
}

/**
 * Format full datetime for detailed display.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - IANA timezone identifier
 * @returns Formatted string like "Monday, January 15, 2024 at 6:30 PM PT"
 *
 * @example
 * formatFullDateTime("2024-01-16T02:30:00Z", "America/Los_Angeles")
 * // Returns: "Monday, January 15, 2024 at 6:30 PM PT"
 */
export function formatFullDateTime(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    const dateStr = formatInUserTimezone(utcString, timezone, "EEEE, MMMM d, yyyy 'at' h:mm a");
    const tzAbbr = getTimezoneAbbr(timezone);
    return `${dateStr} ${tzAbbr}`;
  } catch (error) {
    console.error("Error formatting full datetime:", error);
    return utcString;
  }
}

/**
 * Format date only (no time) for display.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - IANA timezone identifier
 * @returns Formatted string like "January 15, 2024"
 */
export function formatDate(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    return formatInUserTimezone(utcString, timezone, "MMMM d, yyyy");
  } catch (error) {
    console.error("Error formatting date:", error);
    return utcString;
  }
}

/**
 * Format short date for compact display.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - IANA timezone identifier
 * @returns Formatted string like "Jan 15"
 */
export function formatShortDate(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    return formatInUserTimezone(utcString, timezone, "MMM d");
  } catch (error) {
    console.error("Error formatting short date:", error);
    return utcString;
  }
}

/**
 * Convert a local datetime input value to UTC ISO string for API.
 *
 * When a user enters "6:30 PM" in their timezone, this converts it
 * to the equivalent UTC time for storage.
 *
 * @param localDateTimeString - Local datetime string from input (e.g., "2024-01-15T18:30")
 * @param timezone - User's IANA timezone identifier
 * @returns UTC ISO 8601 string (e.g., "2024-01-16T02:30:00.000Z")
 *
 * @example
 * // User in LA enters 6:30 PM on Jan 15
 * localInputToUTC("2024-01-15T18:30", "America/Los_Angeles")
 * // Returns: "2024-01-16T02:30:00.000Z" (next day in UTC)
 */
export function localInputToUTC(
  localDateTimeString: string | null | undefined,
  timezone: string
): string | null {
  if (!localDateTimeString) return null;

  try {
    // Parse the local input as a date in the user's timezone
    const localDate = new Date(localDateTimeString);
    // Convert from the user's timezone to UTC
    const utcDate = fromZonedTime(localDate, timezone);
    return utcDate.toISOString();
  } catch (error) {
    console.error("Error converting local to UTC:", error);
    return null;
  }
}

/**
 * Convert a UTC datetime string to local input format for form editing.
 *
 * When displaying a stored UTC time in a form input, this converts it
 * to the user's local time.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - User's IANA timezone identifier
 * @returns Local datetime string for input (e.g., "2024-01-15T18:30")
 *
 * @example
 * utcToLocalInput("2024-01-16T02:30:00.000Z", "America/Los_Angeles")
 * // Returns: "2024-01-15T18:30"
 */
export function utcToLocalInput(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    const utcDate = parseISO(utcString);
    const zonedDate = toZonedTime(utcDate, timezone);
    // Format for datetime-local input: YYYY-MM-DDTHH:MM
    return format(zonedDate, "yyyy-MM-dd'T'HH:mm");
  } catch (error) {
    console.error("Error converting UTC to local:", error);
    return "";
  }
}

/**
 * Convert a local date input value to UTC ISO string for API.
 *
 * For date-only fields (no time component), sets to start of day in the timezone.
 *
 * @param localDateString - Local date string from input (e.g., "2024-01-15")
 * @param timezone - User's IANA timezone identifier
 * @returns UTC ISO 8601 string
 */
export function localDateToUTC(
  localDateString: string | null | undefined,
  timezone: string
): string | null {
  if (!localDateString) return null;

  try {
    // Parse as start of day in the user's timezone
    const localDate = new Date(`${localDateString}T00:00:00`);
    const utcDate = fromZonedTime(localDate, timezone);
    return utcDate.toISOString();
  } catch (error) {
    console.error("Error converting local date to UTC:", error);
    return null;
  }
}

/**
 * Convert a UTC datetime string to local date input format.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - User's IANA timezone identifier
 * @returns Local date string for input (e.g., "2024-01-15")
 */
export function utcToLocalDate(
  utcString: string | null | undefined,
  timezone: string
): string {
  if (!utcString) return "";

  try {
    const utcDate = parseISO(utcString);
    const zonedDate = toZonedTime(utcDate, timezone);
    return format(zonedDate, "yyyy-MM-dd");
  } catch (error) {
    console.error("Error converting UTC to local date:", error);
    return "";
  }
}

/**
 * Get the current time in the user's timezone.
 *
 * @param timezone - IANA timezone identifier
 * @returns Current Date object in the user's timezone
 */
export function getNowInTimezone(timezone: string): Date {
  return toZonedTime(new Date(), timezone);
}

/**
 * Check if a date string is today in the user's timezone.
 *
 * @param utcString - ISO 8601 UTC datetime string
 * @param timezone - IANA timezone identifier
 * @returns true if the date is today in the user's timezone
 */
export function isToday(
  utcString: string | null | undefined,
  timezone: string
): boolean {
  if (!utcString) return false;

  try {
    const utcDate = parseISO(utcString);
    const zonedDate = toZonedTime(utcDate, timezone);
    const now = getNowInTimezone(timezone);

    return (
      zonedDate.getFullYear() === now.getFullYear() &&
      zonedDate.getMonth() === now.getMonth() &&
      zonedDate.getDate() === now.getDate()
    );
  } catch (error) {
    console.error("Error checking if today:", error);
    return false;
  }
}

/**
 * Get timezone label by value
 */
export function getTimezoneLabel(value: string): string {
  const option = TIMEZONE_OPTIONS.find((opt) => opt.value === value);
  return option?.label || value;
}

File: mvp/backend/app/utils/timezone.py
"""
Centralized timezone utilities for CommonGround.

All datetime handling should use these utilities to ensure consistency:
- Database stores naive UTC datetimes
- API accepts/returns ISO 8601 strings (UTC)
- Frontend handles conversion to/from user's timezone
"""

from datetime import datetime, timezone
from typing import Optional

try:
    from zoneinfo import ZoneInfo  # Python 3.9+
except ImportError:
    from backports.zoneinfo import ZoneInfo  # Fallback for older Python


# Valid US timezone identifiers
VALID_TIMEZONES = [
    "America/New_York",      # Eastern Time
    "America/Chicago",       # Central Time
    "America/Denver",        # Mountain Time
    "America/Los_Angeles",   # Pacific Time
    "America/Anchorage",     # Alaska Time
    "Pacific/Honolulu",      # Hawaii Time
]

DEFAULT_TIMEZONE = "America/Los_Angeles"


def utc_now() -> datetime:
    """
    Get current UTC time as naive datetime.

    Use this instead of datetime.utcnow() for consistency.
    Note: datetime.utcnow() is deprecated in Python 3.12+
    """
    return datetime.now(timezone.utc).replace(tzinfo=None)


def strip_tz(dt: Optional[datetime]) -> Optional[datetime]:
    """
    Convert timezone-aware datetime to naive UTC datetime.

    Use when receiving datetimes from API requests that may have
    timezone information attached.

    Args:
        dt: A datetime object (may or may not have tzinfo)

    Returns:
        Naive datetime in UTC, or None if input was None
    """
    if dt is None:
        return None
    if dt.tzinfo is not None:
        utc_dt = dt.astimezone(timezone.utc)
        return utc_dt.replace(tzinfo=None)
    return dt


def to_utc_naive(dt: datetime, from_tz: str) -> datetime:
    """
    Convert a naive datetime from a specific timezone to naive UTC.

    Use when a user enters a time in their local timezone and
    you need to store it as UTC.

    Args:
        dt: Naive datetime representing local time
        from_tz: IANA timezone identifier (e.g., "America/Los_Angeles")

    Returns:
        Naive datetime in UTC

    Example:
        # User in LA enters 6:30 PM on Jan 15
        la_time = datetime(2024, 1, 15, 18, 30)
        utc_time = to_utc_naive(la_time, "America/Los_Angeles")
        # Result: datetime(2024, 1, 16, 2, 30) - next day in UTC
    """
    if dt.tzinfo is not None:
        raise ValueError("Expected naive datetime, got timezone-aware datetime")

    tz = ZoneInfo(from_tz)
    local_dt = dt.replace(tzinfo=tz)
    utc_dt = local_dt.astimezone(timezone.utc)
    return utc_dt.replace(tzinfo=None)


def from_utc_naive(dt: datetime, to_tz: str) -> datetime:
    """
    Convert naive UTC datetime to a specific timezone (still naive).

    Use for display purposes when you need the local time value.

    Args:
        dt: Naive datetime in UTC
        to_tz: IANA timezone identifier (e.g., "America/New_York")

    Returns:
        Naive datetime in the target timezone

    Example:
        # UTC time is 2:30 AM on Jan 16
        utc_time = datetime(2024, 1, 16, 2, 30)
        ny_time = from_utc_naive(utc_time, "America/New_York")
        # Result: datetime(2024, 1, 15, 21, 30) - 9:30 PM previous day
    """
    if dt.tzinfo is not None:
        raise ValueError("Expected naive datetime, got timezone-aware datetime")

    utc_dt = dt.replace(tzinfo=timezone.utc)
    tz = ZoneInfo(to_tz)
    local_dt = utc_dt.astimezone(tz)
    return local_dt.replace(tzinfo=None)


def validate_timezone(tz: str) -> bool:
    """
    Check if timezone identifier is in our supported list.

    Args:
        tz: IANA timezone identifier

    Returns:
        True if valid, False otherwise
    """
    return tz in VALID_TIMEZONES


def format_for_api(dt: Optional[datetime]) -> Optional[str]:
    """
    Format datetime for API response.

    Returns ISO 8601 string with Z suffix to indicate UTC.

    Args:
        dt: Naive datetime (assumed UTC)

    Returns:
        ISO 8601 string like "2024-01-15T18:30:00Z", or None
    """
    if dt is None:
        return None
    return dt.isoformat() + "Z"


def parse_from_api(dt_string: Optional[str]) -> Optional[datetime]:
    """
    Parse datetime string from API request.

    Handles ISO 8601 strings with or without timezone info.
    Always returns naive UTC datetime.

    Args:
        dt_string: ISO 8601 datetime string

    Returns:
        Naive datetime in UTC, or None
    """
    if dt_string is None:
        return None

    # Parse the string - fromisoformat handles Z and +00:00
    dt_string = dt_string.replace("Z", "+00:00")
    dt = datetime.fromisoformat(dt_string)

    # Convert to UTC and strip timezone
    return strip_tz(dt)
